statistics/update_test.go

```diff
@@ -918,3 +919,107 @@ func (s *testStatsUpdateSuite) TestLogDetailedInfo(c *C) {
 		c.Assert(s.hook.results, Equals, t.result)
 	}
 }
+
+func (s *testStatsUpdateSuite) TestNeedAnalyzeTable(c *C) {
+	columns := map[int64]*statistics.Column{}
+	columns[1] = &statistics.Column{Count: 1}
+	tests := []struct {
+		tbl    *statistics.Table
+		ratio  float64
+		limit  time.Duration
+		start  string
+		end    string
+		now    string
+		result bool
+	}{
+		// table never analyzed and has reach the limit
+		{
+			tbl:    &statistics.Table{Version: oracle.EncodeTSO(oracle.GetPhysical(time.Now()))},
+			limit:  0,
+			ratio:  0,
+			start:  "00:00 CST",
+			end:    "00:01 CST",
+			now:    "00:00 CST",
+			result: true,
+		},
+		// table never analyzed but has not reach the limit
+		{
+			tbl:    &statistics.Table{Version: oracle.EncodeTSO(oracle.GetPhysical(time.Now()))},
+			limit:  time.Hour,
+			ratio:  0,
+			start:  "00:00 CST",
+			end:    "00:01 CST",
+			now:    "00:00 CST",
+			result: false,
+		},
+		// table already analyzed but auto analyze is disabled
+		{
+			tbl:    &statistics.Table{HistColl: statistics.HistColl{Columns: columns, ModifyCount: 1, Count: 1}},
+			limit:  0,
+			ratio:  0,
+			start:  "00:00 CST",
+			end:    "00:01 CST",
+			now:    "00:00 CST",
+			result: false,
+		},
+		// table already analyzed and but modify count is small
+		{
+			tbl:    &statistics.Table{HistColl: statistics.HistColl{Columns: columns, ModifyCount: 0, Count: 1}},
+			limit:  0,
+			ratio:  0.3,
+			start:  "00:00 CST",
+			end:    "00:01 CST",
+			now:    "00:00 CST",
+			result: false,
+		},
+		// table already analyzed and but not within time period
+		{
+			tbl:    &statistics.Table{HistColl: statistics.HistColl{Columns: columns, ModifyCount: 1, Count: 1}},
+			limit:  0,
+			ratio:  0.3,
+			start:  "00:00 CST",
+			end:    "00:01 CST",
+			now:    "00:02 CST",
+			result: false,
+		},
+		// table already analyzed and but not within time period
```

Comment:
s/table already analyzed and but not within time period/table is already analyzed and but not within time period

and for the rest of the comment, you also need fix them.